src/werkzeug/routing/map.py

```diff
@@ -38,907 +38,916 @@
     from ..wrappers.request import Request
 
 
-class Map:
```

Comment:
This diff is large as I've moved the MapAdapter definition to before the Map definition, so that the latter can use it.